File: backend/app/routes/routine.py
from fastapi import APIRouter, Depends, HTTPException, Query

from app.auth import get_current_user_id
from app.models.schemas import Period, RoutineSaveRequest, UserRoutine
from app.services.routine_service import build_routine, get_routine, save_routine

router = APIRouter()


@router.get("", response_model=UserRoutine)
def fetch_routine(
    period: Period,
    user_id: str = Depends(get_current_user_id),
) -> UserRoutine:
    routine = get_routine(user_id, period)
    if not routine:
        return UserRoutine(period=period, steps=[], status="INCOMPLETE")
    return routine


@router.post("/build", response_model=UserRoutine)
def auto_build_routine(
    period: Period,
    user_id: str = Depends(get_current_user_id),
) -> UserRoutine:
    try:
        return build_routine(user_id, period)
    except Exception as exc:
        raise HTTPException(status_code=500, detail=str(exc)) from exc


@router.post("", response_model=UserRoutine)
def save_user_routine(
    body: RoutineSaveRequest,
    user_id: str = Depends(get_current_user_id),
) -> UserRoutine:
    try:
        return save_routine(user_id, body.period, body.steps)
    except Exception as exc:
        raise HTTPException(status_code=500, detail=str(exc)) from exc

File: mobile/src/services/routine.ts
import { apiFetch } from './api';

export type Period = 'AM' | 'PM';
export type RoutineStatus = 'READY' | 'INCOMPLETE';

export interface RoutineStep {
  order: number;
  product_id: string;
  product_name: string;
  brand?: string | null;
  category: string;
  reason: string;
}

export interface UserRoutine {
  period: Period;
  steps: RoutineStep[];
  status: RoutineStatus;
  updated_at?: string | null;
}

export async function getRoutine(period: Period): Promise<UserRoutine> {
  return apiFetch<UserRoutine>(`/routine?period=${period}`);
}

export async function buildRoutine(period: Period): Promise<UserRoutine> {
  return apiFetch<UserRoutine>(`/routine/build?period=${period}`, { method: 'POST' });
}

export async function saveRoutine(period: Period, steps: RoutineStep[]): Promise<UserRoutine> {
  return apiFetch<UserRoutine>('/routine', {
    method: 'POST',
    body: JSON.stringify({ period, steps }),
  });
}
